### PR TITLE
Support KL loss

### PR DESCRIPTION
**Pull request by zqhuang211**
This PR implements the (response) KL loss described in the BLSP-KD paper: https://arxiv.org/abs/2405.19041

### CODE REVIEW

**Comment by zqhuang211** [comment]
I am running model training to verify training works properly.

File: ultravox/model/data_processing.py [hunk 1]
```diff
@@ -91,8 +110,33 @@ def _process(self, sample: datasets.VoiceSample) -> Dict[str, Any]:
             )["input_ids"].shape[-1]
             labels[:input_text_len] = -100
 
+            if self.include_alt_input:
+                alt_input_text = input_text.replace(
```
Comment (juberti) [inline ultravox/model/data_processing.py:114]: Can this be simplified? The assistant message should have the same length for both input types, so can we simply do `alt_input_text_len = len(alt_text) - (len(text) - len(input_text))`?

**Comment by zqhuang211** [comment]
Model training finished as expected.

File: ultravox/model/ultravox_model.py [hunk 2]
```diff
@@ -75,6 +78,10 @@ def get_decoder(self):
     def tie_weights(self):
         return self.language_model.tie_weights()
 
+    def set_loss_config(self, loss_config: Optional[LossConfig] = None):
```
Comment (juberti) [inline ultravox/model/ultravox_model.py:81]: Why `Optional`?

File: ultravox/model/ultravox_model.py [hunk 3]
```diff
@@ -162,8 +173,40 @@ def forward(
             past_key_values=past_key_values,
             **kwargs,
         )
-
-        return lm_output
+        if self.training:
+            if self.loss_config.loss_function == LossFunction.CrossEntropy:
+                return lm_output
+            elif self.loss_config.loss_function == LossFunction.KL_Divergence:
+                # compute the teacher (text-only) model's distribution
```
Comment (juberti) [inline ultravox/model/ultravox_model.py:180]: suggest hoisting this into a `self._compute_kl_loss()` function to try to keep the core `forward()` logic clear

File: ultravox/model/ultravox_config.py [hunk 4]
```diff
@@ -20,6 +21,20 @@ class LoraConfigSimplified:
...
+class LossFunction(str, Enum):
+    CrossEntropy = "ce"
+    KL_Divergence = "kl"
+
+
+@dataclasses.dataclass
+class LossConfig:
+    loss_function: LossFunction = LossFunction.KL_Divergence
+    kl_temperature: float = 2.0
+
```
Comment (juberti) [inline ultravox/model/ultravox_config.py:33]: suggest making this a `@property`

File: ultravox/training/train.py [hunk 5]
```diff
@@ -155,6 +159,8 @@ def main() -> None:
                 )
 
     model.print_trainable_parameters()
+    # loss_config needs to be passed separately just for model training
+    model.set_loss_config(args.loss_config)
```
Comment (juberti) [inline ultravox/training/train.py:163]: suggest moving this before line 128 so all model init happens in one place. Also, suggest only calling `set_loss_config` if `args.loss_config` isn't None, to avoid the need for `Optional`

**Review by juberti — COMMENTED** [review]
Thanks for splitting this out, mostly minor comments.

File: mcloud.yaml [hunk 6]
```diff
@@ -13,5 +13,5 @@ command: >-
   cd ultravox && poetry install --no-dev && poetry run torchrun --nproc_per_node=8 -m ultravox.training.train $TRAIN_ARGS
 env_variables:
   MLFLOW_TRACKING_URI: databricks
-  UV_BRANCH: main
-  TRAIN_ARGS: --config_path ultravox/training/configs/llama3_whisper.yaml
+  UV_BRANCH: zhuang/add-kl-loss
```
[the remark this answers is not in the document]
Comment (zqhuang211) [reply mcloud.yaml]: Reverted after training run passed.

File: ultravox/data/datasets.py [hunk 7]
```diff
@@ -69,9 +69,27 @@
 
 @dataclasses.dataclass
 class DataCollatorForSeq2SeqWithAudio(transformers.DataCollatorForSeq2Seq):
+    # when enabled, the alt_input_ids, alt_attention_mask, and alt_labels fields are used for computing the KL loss in UltravoxModel
+    include_alt_input: bool = False
```
[the remark this answers is not in the document]
Comment (zqhuang211) [reply ultravox/data/datasets.py:73]: We need to prepare alternative versions of the fields `input_ids`, `attention_mask`, and `labels` in a batch for text input. I struggled with finding the right name for the flag. I wanted to emphasize `alt_` to match the `alt_*` fields. However, `input_alt_input` does not capture the `alt_attention_mask` and `alt_labels` fields. Similarly, `include_text_input` does not account for the `alt_` prefix or the other fields.

Perhaps we could use `include_alt_fields`?

File: ultravox/model/data_processing.py [hunk 8]
```diff
@@ -45,6 +49,9 @@ def _process(self, sample: datasets.VoiceSample) -> Dict[str, Any]:
             sample.messages, tokenize=False
         )
 
+        if self.include_alt_input:
+            alt_text = text.replace("<|audio|>", sample.audio_transcript or "")
```
[the remark this answers is not in the document]
Comment (zqhuang211) [reply ultravox/model/data_processing.py:53]: The `_get_sample()` method in `datasets.py` would be a more appropriate place to handle this logic. I think it’s acceptable to support empty audio transcripts in this context, as an audio file might simply contain silence or noise.

File: ultravox/model/data_processing.py [hunk 9]
```diff
@@ -65,9 +72,21 @@ def _process(self, sample: datasets.VoiceSample) -> Dict[str, Any]:
...
+                audio=None,
+                return_tensors="pt",
+            )
+            alt_input_ids = alt_inputs["input_ids"].squeeze_(0)
+            alt_inputs["attention_mask"].squeeze_(0)
+
         # No need to shift the labels as the model does it internally
         labels = input_ids.clone()
 
+        if self.include_alt_input:
```
[the remark this answers is not in the document]
Comment (zqhuang211) [reply ultravox/model/data_processing.py:87]: I debated this. Keep them separate is easier for implementing the parallels, but I agree it is probably better for readability. To be updated.

File: ultravox/model/data_processing.py (hunk 1 above)
Comment (zqhuang211) [reply ultravox/model/data_processing.py:114]: I think you are right, though the computation should be based on tokens, not text.

`alt_input_text_len = input_text_len + len(alt_input_ids) - len(input_ids)`

File: ultravox/model/ultravox_model.py (hunk 2 above)
Comment (zqhuang211) [reply ultravox/model/ultravox_model.py:81]: This is because `loss_config` defaults to None in `config_base.py`, but it shouldn't be None during training. I will update `config_base.py` to make `loss_confg` to have a non-None default value, then we can get rid of Optional here.

File: ultravox/data/datasets.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (juberti) [reply ultravox/data/datasets.py:73]: Isn't the `alt` prefix something that we added? That is, could we just make a global change to use `text` or some similar prefix? 

Or, if we want to keep `alt`, perhaps `include_alt_modality` suggesting that we're trying to send multiple modalities to the model for the same input.

File: ultravox/model/data_processing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (juberti) [reply ultravox/model/data_processing.py:53]: My concern was more that if there was some logic error that led to a `None` transcript, we should probably handle that rather than silently inserting an empty transcript. Even if the audio is silent, there should still be a valid "" transcript.

File: ultravox/model/data_processing.py (hunk 1 above)
Comment (juberti) [reply ultravox/model/data_processing.py:114]: ah yeah, good catch. maybe `alt_input_token_len` then to avoid future confusion.

File: ultravox/model/ultravox_model.py (hunk 3 above)
Comment (zqhuang211) [reply ultravox/model/ultravox_model.py:180]: Makes sense. Will update

File: ultravox/model/data_processing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zqhuang211) [reply ultravox/model/data_processing.py:53]: Agreed.

File: ultravox/model/data_processing.py (hunk 1 above)
Comment (zqhuang211) [reply ultravox/model/data_processing.py:114]: right, i will update `input_text_len` to `input_token_len` as well.

File: ultravox/model/ultravox_config.py (hunk 4 above)
Comment (zqhuang211) [reply ultravox/model/ultravox_config.py:33]: Sounds like the right thing to do. Please suggest how to do it.

File: ultravox/training/train.py (hunk 5 above)
Comment (zqhuang211) [reply ultravox/training/train.py:163]: Done. args.loss_config is updated to always have the LossConfig type.

File: ultravox/model/ultravox_config.py (hunk 4 above)
Comment (juberti) [reply ultravox/model/ultravox_config.py:33]: This should be sufficient:

```suggestion
@property
```

File: ultravox/model/data_processing.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (zqhuang211) [reply ultravox/model/data_processing.py]: I had to keep the `or ""` statement for now, otherwise some samples cause training failure. Will revisit this later as we start to clean up `datasets.py`

File: ultravox/model/ultravox_model.py (hunk 2 above)
Comment (zqhuang211) [reply ultravox/model/ultravox_model.py]: It turns out that I have to use Optional[LossConfig] in config_base.py, otherwise it will not pass `just check`.  I updated `train.py` to only invoke `set_loss_config` when `args.loss_config is not None`

**Comment by zqhuang211** [comment]
@juberti please review the latest version when you have a chance.

**Review by juberti — APPROVED** [review]

File: ultravox/model/ultravox_model.py [hunk 10]
```diff
@@ -162,8 +206,25 @@ def forward(
             past_key_values=past_key_values,
             **kwargs,
         )
-
-        return lm_output
+        if self.training:
```
Comment (farzadab) [inline ultravox/model/ultravox_model.py:209]: Hmm, why gate this with `self.training`?
Do you explicitly want the loss to be different in training vs validation?

Is is because you want the validation loss to be comparable to other runs?

File: ultravox/training/configs/llama3_whisper_kd.yaml [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+
+# Temporarily remove heysquad_human from val_sets as it causes the training to fail.
+val_sets: ["anyinstruct", "soda", "peoplespeech"]
+
+# Load continuation training data using GenericVoiceDataset.
+# Select 100,000 samples from both clean and other subsets.
+# Mix clean and other subsets in a 2:1 ratio for batch creation.
+# Total training steps: 4000, with each step 4x8=32 samples.
+batch_size: 4
+max_steps: 1000
```
Comment (farzadab) [inline ultravox/training/configs/llama3_whisper_kd.yaml:21]: Doesn't seem like the comments are reflecting the config itself? E.g. `max_steps` is 1000 not 4000. There aren't 100K samples being used here, etc. So I suggest removing them.

File: mcloud.yaml [hunk 12]
```diff
@@ -13,5 +13,6 @@ command: >-
   cd ultravox && poetry install --no-dev && poetry run torchrun --nproc_per_node=8 -m ultravox.training.train $TRAIN_ARGS
 env_variables:
   MLFLOW_TRACKING_URI: databricks
+  # temperary for training verfication
   UV_BRANCH: main
-  TRAIN_ARGS: --config_path ultravox/training/configs/llama3_whisper.yaml
+  TRAIN_ARGS: --config_path ultravox/training/configs/llama3_whisper_kd.yaml
```
Comment (farzadab) [inline mcloud.yaml:18]: Nit: If this is temporary, then make sure to remove it from PR :?

**Review by farzadab — APPROVED** [review]

File: ultravox/data/datasets.py [hunk 13]
```diff
@@ -433,15 +451,13 @@ class AnyInstructDataset(VoiceDataset):
...
-        # The last 7 samples are missing audio files, so we exclude them.
-        NUM_SAMPLES = 108193 - 7
         super().__init__(args)
         dataset = datasets.load_dataset(
             "json",
             "anyinstruct",
             data_files="https://huggingface.co/datasets/fnlp/AnyInstruct/resolve/main/speech_conv/metadata.jsonl",
             split="train",
-        ).select(range(NUM_SAMPLES))
+        )
```
[the remark this answers is not in the document]
Comment (juberti) [reply ultravox/data/datasets.py:460]: Apparently the dataset has been updated? Although we should verify that the speech data is still in sync with our local copy.

File: ultravox/model/data_processing.py [hunk 14]
```diff
@@ -84,15 +88,40 @@ def _process(self, sample: datasets.VoiceSample) -> Dict[str, Any]:
...
+            alt_labels = alt_input_ids.clone()
+            if not self.train_on_inputs:
+                alt_input_token_len = (
+                    input_token_len + len(alt_input_ids) - len(input_ids)
+                )
+                alt_labels[:alt_input_token_len] = -100
+
+            inputs["alt_input_ids"] = alt_input_ids
+            inputs["alt_attention_mask"] = alt_inputs["attention_mask"]
+            inputs["alt_labels"] = alt_labels
```
[the remark this answers is not in the document]
Comment (zqhuang211) [reply ultravox/model/data_processing.py:120]: We will look into this later.

File: ultravox/data/datasets.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (zqhuang211) [reply ultravox/data/datasets.py:460]: The dataset on huggingface was updated
